core/dbt/adapters/base/impl.py

```diff
@@ -1134,6 +1134,14 @@ def get_rows_different_sql(
 
         return sql
 
+    @available.parse_none
+    def submit_python_job(self, parsed_model: dict, compiled_code: str):
+        raise NotImplementedException("`submit_python_job` is not implemented for this adapter!")
+
+    def valid_python_model_materialization(self):
+        """The valid materialization an adapter should have for python model"""
+        return ["table"]
```

Comment:
Maybe we shouldn't have a default here, but then this makes testing harder

Reply:
Hmm. Would a check like this one prohibit users from defining their own custom materializations that support Python (not named `table`)? Given all we've been discussing about guardrails, this idea isn't so wild to me — but it does go against the flexibility we've allowed user-space code thus far.

I feel like the right pattern here might look like:
- A mapping, stored on the adapter, of `{language: [supported, materializations]}`. (For `sql`, that can just be "all" of them.)
- Each materialization declaring which languages it supports — either as information we can parse statically (tricky, perhaps ideal), _or_ a validation check at runtime — very easy, and still better than status quo, where dbt tries to wrap Python code in `create view as`

That would require a teensy bit of net-new code in the `view` materialization, for those adapters which support Python model code for `table` + `incremental`

Reply:
I see, my original idea was that this is a function to overwrite in adapter's `impl.py`.(I did that for dbt-snowflake and dbt-spark). And since we are checking it at parsing time we won't need to actually run any model before an error is raised. 

But this does prohibit users who can't modify the python code in adapter from adding new materialization.

Happy to move this to a simple runtime check.